src/components/implementation/tests/unit_defcompinfo/unit_defcompinfo.c

```diff
@@ -0,0 +1,238 @@
+#define SPIN() do { while (1) ; } while (0)
+
+#define CHILD_COMP_COUNT   2
+#define CHILD_UNTYPED_SIZE (1<<24)
+#define CHILD_SCHED_ID     0
+#define CHILD_SCHED_CYCS   TCAP_RES_INF 
+#define CHILD_SCHED_PRIO   TCAP_PRIO_MAX
+
+static void
+cos_llprint(char *s, int len)
+{ call_cap(PRINT_CAP_TEMP, (int)s, len, 0, 0); }
+
+int
+prints(char *s)
+{
+	int len = strlen(s);
+
+	cos_llprint(s, len);
+
+	return len;
+}
+
+int __attribute__((format(printf,1,2)))
+printc(char *fmt, ...)
+{
+	  char s[128];
+	  va_list arg_ptr;
+	  int ret, len = 128;
+
+	  va_start(arg_ptr, fmt);
+	  ret = vsnprintf(s, len, fmt, arg_ptr);
+	  va_end(arg_ptr);
+	  cos_llprint(s, ret);
+
+	  return ret;
+}
+
+int is_booter = 1;
+extern vaddr_t cos_upcall_entry;
+struct cos_defcompinfo child_defci[CHILD_COMP_COUNT];
+static cycles_t cycs_per_usec;
+
+#define TEST_AEPS_N   3
+#define TEST_AEP_CYCS 400000
+#define TEST_AEP_PRIO TCAP_PRIO_MAX
+struct cos_aep_info test_aep[TEST_AEPS_N];
+
+static void
+aep_thd_fn(arcvcap_t rcv, void *data)
+{
+	printc("\tSwitched to aep %d\n", (int)data);
+	while (1) {
+		cos_rcv(rcv);
+	}
+}
+
+static void
+test_aeps(void)
+{
+	int                     i, ret;
+	int                     blocked;
+	cycles_t                cycs;
+	thdid_t                 tid;
+	struct cos_compinfo    *ci    = cos_compinfo_get(cos_defcompinfo_curr_get());
+
+	memset(&test_aep, 0, sizeof(struct cos_aep_info) * TEST_AEPS_N);
+
+	printc("Test creating AEPS\n");
+	for (i = 0 ; i < TEST_AEPS_N ; i ++) {
+		asndcap_t snd;
+
+		printc("\tCreating AEP [%d]\n", i);
+		ret = cos_aep_tcap_alloc(&(test_aep[i]), BOOT_CAPTBL_SELF_INITTCAP_BASE, aep_thd_fn, (void *)i);
+		assert(ret == 0);
+
+		snd = cos_asnd_alloc(ci, test_aep[i].rcv, ci->captbl_cap);
+		assert(snd);
+			
+		ret = cos_tcap_delegate(snd, BOOT_CAPTBL_SELF_INITTCAP_BASE, TEST_AEP_CYCS, TEST_AEP_PRIO, TCAP_DELEG_YIELD);
+		assert(ret == 0);
+
+		while (cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_BASE, &tid, &blocked, &cycs));
+	}
+
+	printc("Done.\n");
+}
+
+static void
+test_childcomps(void)
+{
+	int id, ret;
+
+	printc("Test switching to new components\n");
+	for (id = 0 ; id < CHILD_COMP_COUNT ; id ++) {
+		int      blocked;
+		cycles_t cycs;
+		thdid_t  tid;
+
+		while (cos_sched_rcv(BOOT_CAPTBL_SELF_INITRCV_BASE, &tid, &blocked, &cycs));
+		printc("\tSwitching to [%d] component\n", id);
+		if (id == CHILD_SCHED_ID) {
+			ret = cos_switch(child_defci[id].sched_aep.thd, child_defci[id].sched_aep.tc, CHILD_SCHED_PRIO, TCAP_TIME_NIL, 
+					BOOT_CAPTBL_SELF_INITRCV_BASE, cos_sched_sync());
+			assert(ret == 0);
+		} else {
+			cycles_t    now;
+			tcap_time_t timer;
+
+			rdtscll(now);
+			timer = tcap_cyc2time(now + 100 * cycs_per_usec);
+
+			ret = cos_defswitch(child_defci[id].sched_aep.thd, timer, CHILD_SCHED_PRIO, cos_sched_sync());
+			assert(ret == 0);
+		}
+	}
+
+	printc("Done.\n");
+}
+
+void
+cos_init(void)
+{
+	int cycs;
+
+	while (!(cycs = cos_hw_cycles_per_usec(BOOT_CAPTBL_SELF_INITHW_BASE))) ;
```

Comment:
Okay, I'll do that. I had another idea to call this from upcall entry function, https://github.com/phanikishoreg/composite/commit/3e5c50eeec82c2481c239dbcbd216c1879afd9d0, if that sounds good, I'll incorporate that here.